mopidy/files/library.py

```diff
@@ -0,0 +1,155 @@
+import operator
+import os
+import stat
+import sys
+import urllib2
+
+
+from mopidy import backend, exceptions, models
+from mopidy.audio import scan, utils
+from mopidy.internal import path
+
+logger = logging.getLogger(__name__)
+
+
+class FilesLibraryProvider(backend.LibraryProvider):
+    """Library for browsing local files."""
+
+    @property
+    def root_directory(self):
+        if not self._media_dirs:
+            return None
+        elif len(self._media_dirs) == 1:
+            localpath = self._media_dirs[0]['path']
+            uri = path.path_to_uri(localpath)
+        else:
+            uri = u'file:root'
+        return models.Ref.directory(name='Files', uri=uri)
+
+    def __init__(self, backend, config):
+        super(FilesLibraryProvider, self).__init__(backend)
+        self._media_dirs = []
+        # import pdb; pdb.set_trace()
+        for entry in config['files']['media_dir']:
+            media_dir = {}
+            media_dict = entry.split(':')
+            local_path = path.expand_path(
+                media_dict[0].encode(sys.getfilesystemencoding()))
+            st = os.stat(local_path)
+            if not stat.S_ISDIR(st.st_mode):
+                logger.warn(u'%s is not a directory' % local_path)
+                continue
+            media_dir['path'] = local_path
+            if len(media_dict) == 2:
+                media_dir['name'] = media_dict[1]
+            else:
+                media_dir['name'] = media_dict[0].replace(os.sep, '+')
+            self._media_dirs.append(media_dir)
+        logger.debug(self._media_dirs)
+        self._follow_symlinks = config['files']['follow_symlinks']
+        self._show_hidden = config['files']['show_hidden']
+        self._scanner = scan.Scanner(
+            timeout=config['files']['metadata_timeout'])
+
+    def browse(self, uri, encoding=sys.getfilesystemencoding()):
+        logger.debug(u'browse called with uri %s' % uri)
+        # import pdb; pdb.set_trace()
+        result = []
+        localpath = path.uri_to_path(uri)
+        if localpath == 'root':
+            result = self._show_media_dirs()
+        else:
+            if not self._is_in_basedir(localpath):
+                logger.warn(u'Not in basedir: %s' % localpath)
+                return []
+            for name in os.listdir(localpath):
+                child = os.path.join(localpath, name)
+                uri = path.path_to_uri(child)
+                name = name.decode('ascii', 'ignore')
+                if self._follow_symlinks:
+                    st = os.stat(child)
+                else:
+                    st = os.lstat(child)
+                if not self._show_hidden and name.startswith(b'.'):
+                    continue
+                elif stat.S_ISDIR(st.st_mode):
+                    result.append(models.Ref.directory(name=name, uri=uri))
+                elif stat.S_ISREG(st.st_mode) and self._check_audiofile(uri):
+                    # if self._is_playlist(child):
```

Comment:
There's also the subject of the key-value list (aka dictionary) config.
I've seen that mopidy-local-sqlite has a very similar thing but does it
slightly differently (with spaces rather than ':' separator characters. Do
you think it's worth trying to standardise this sort of thing? Through a
new config type or just a convention?
On 30 Jun 2015 20:22, "Thomas Adamcik" <notifications@github.com> wrote:

> In mopidy/files/library.py
> <https://github.com/mopidy/mopidy/pull/1207#discussion_r33612833>:
>
> > +                logger.warn(u'Not in basedir: %s' % localpath)
> > +                return []
> > +            for name in os.listdir(localpath):
> > +                child = os.path.join(localpath, name)
> > +                uri = path.path_to_uri(child)
> > +                name = name.decode('ascii', 'ignore')
> > +                if self._follow_symlinks:
> > +                    st = os.stat(child)
> > +                else:
> > +                    st = os.lstat(child)
> > +                if not self._show_hidden and name.startswith(b'.'):
> > +                    continue
> > +                elif stat.S_ISDIR(st.st_mode):
> > +                    result.append(models.Ref.directory(name=name, uri=uri))
> > +                elif stat.S_ISREG(st.st_mode) and self._check_audiofile(uri):
> > +                    # if self._is_playlist(child):
>
> Remove the commented out code and/or add a TODO with what is missing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mopidy/mopidy/pull/1207/files#r33612833>.
>